provider/pkg/docs-gen/examples/generate.go

```diff
@@ -0,0 +1,214 @@
+
+func main() {
+	if len(os.Args) < 3 {
+		fmt.Fprintf(os.Stdout, "Usage: %s <yaml source dir path> <markdown destination path>\n", os.Args[0])
+		os.Exit(1)
+	}
+	yamlPath := os.Args[1]
+	mdPath := os.Args[2]
+
+	if !filepath.IsAbs(yamlPath) {
+		cwd, err := os.Getwd()
+		contract.AssertNoError(err)
+		yamlPath = filepath.Join(cwd, yamlPath)
+	}
+
+	fileInfo, err := os.Lstat(mdPath)
+	if err != nil && os.IsNotExist(err) {
+		if err := os.MkdirAll(mdPath, 0600); err != nil {
+			panic(err)
+		}
+	}
+
+	if !fileInfo.IsDir() {
+		fmt.Fprintf(os.Stderr, "Expect markdown destination %q to be a directory\n", mdPath)
+		os.Exit(1)
+	}
+
+	yamlFiles, err := os.ReadDir(yamlPath)
+	if err != nil {
+		panic(err)
+	}
+	for _, yamlFile := range yamlFiles {
+		if err := processYaml(filepath.Join(yamlPath, yamlFile.Name()), mdPath); err != nil {
+			fmt.Fprintf(os.Stderr, "%+v", err)
+			os.Exit(1)
+		}
+	}
+}
+
+func markdownExamples(examples []string) string {
+	s := "{{% examples %}}\n## Example Usage\n"
+	for _, example := range examples {
+		s += example
+	}
+	s += "{{% /examples %}}\n"
+	return s
+}
+
+func markdownExample(description string,
+	typescript string,
+	python string,
+	csharp string,
+	golang string,
+	yaml string,
+	java string) string {
+
+	return fmt.Sprintf("{{%% example %%}}\n### %s\n\n"+
+		"```typescript\n%s```\n"+
+		"```python\n%s```\n"+
+		"```csharp\n%s```\n"+
+		"```go\n%s```\n"+
+		"```yaml\n%s```\n"+
+		"```java\n%s```\n"+
+		"{{%% /example %%}}\n",
+		description, typescript, python, csharp, golang, yaml, java)
+}
+
+func processYaml(path string, mdDir string) error {
+	yamlFile, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+
+	base := filepath.Base(path)
+	md := strings.NewReplacer(".yaml", ".md", ".yml", ".md").Replace(base)
+
+	defer contract.IgnoreClose(yamlFile)
+	decoder := yaml.NewDecoder(yamlFile)
+	exampleStrings := []string{}
+	for {
+		example := map[string]interface{}{}
+		err := decoder.Decode(&example)
+		if err == io.EOF {
+			break
+		}
+
+		description := example["description"].(string)
+		dir, err := os.MkdirTemp("", "")
+		if err != nil {
+			return err
+		}
+
+		defer func() {
+			contract.IgnoreError(os.RemoveAll(dir))
+		}()
+
+		src, err := os.OpenFile(filepath.Join(dir, "Pulumi.yaml"), os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0755)
+		if err != nil {
+			return err
+		}
+
+		if err = yaml.NewEncoder(src).Encode(example); err != nil {
+			return err
+		}
+		contract.AssertNoError(src.Close())
+
+		cmd := exec.Command("pulumi", "convert", "--language", "typescript", "--out",
+			filepath.Join(dir, "example-nodejs"), "--generate-only")
+		cmd.Stderr = os.Stderr
+		cmd.Stdout = os.Stdout
+		fmt.Println(dir)
+		cmd.Dir = dir
+		if err = cmd.Run(); err != nil {
+			_, _ = fmt.Fprintf(os.Stderr, "convert nodejs failed, ignoring: %+v", err)
+		}
+		content, err := os.ReadFile(filepath.Join(dir, "example-nodejs", "index.ts"))
+		if err != nil {
+			return err
+		}
+		typescript := string(content)
```

Comment:
This exact pattern keeps appearing. Let's put it in a a lambda.
```suggestion
		convert := func(language, dir, outputFile string) (string, error) {
			exDir := filepath.Join(dir, "example-"+dir)
			cmd := exec.Command("pulumi", "convert", "--language", language, "--out",
				exDir, "--generate-only")
			cmd.Stderr = os.Stderr
			cmd.Stdout = os.Stdout
			cmd.Dir = dir
			if err = cmd.Run(); err != nil {
				_, _ = fmt.Fprintf(os.Stderr, "convert %s failed, language ignoring: %+v", err)
			}
			content, err := os.ReadFile(exDir, "index.ts"))
			if err != nil {
				return "", err
			}
			return string(content), nil
		}
		
		typescript, err := convert("typescript", "nodejs", "index.ts")
		if err != nil { return err }
```

We can then do python via `python, err := convert("python", "py", "__main__.py")` instead of repeating this block.